okhttp-tests/src/test/java/okhttp3/internal/tls/ClientAuthTest.java

```diff
@@ -0,0 +1,176 @@
+        .serialNumber("3")
+        .commonName(server.getHostName())
+        .build();
+
+    clientCert = new HeldCertificate.Builder()
+        .issuedBy(intermediateCa)
+        .serialNumber("4")
+        .commonName("Jethro Willis")
+        .build();
+  }
+
+  @Test public void clientAuth() throws Exception {
+    OkHttpClient client = buildClient(clientCert, intermediateCa);
+
+    SSLSocketFactory socketFactory = buildServerSslSocketFactory(ClientAuth.WANTS);
+
+    server.useHttps(socketFactory, false);
+    server.enqueue(new MockResponse().setBody("abc"));
+
+    Call call1 = client.newCall(new Request.Builder().url(server.url("/")).build());
+    Response response1 = call1.execute();
+    assertEquals(new X500Principal("CN=localhost"), response1.handshake().peerPrincipal());
+    assertEquals(new X500Principal("CN=Jethro Willis"), response1.handshake().localPrincipal());
+    assertEquals("abc", response1.body().string());
+  }
+
+  @Test public void missingClientAuthSkippedForWantsOnly() throws Exception {
+    OkHttpClient client = buildClient(null, intermediateCa);
+
+    SSLSocketFactory socketFactory = buildServerSslSocketFactory(ClientAuth.WANTS);
+
+    server.useHttps(socketFactory, false);
+    server.enqueue(new MockResponse().setBody("abc"));
+
+    Call call1 = client.newCall(new Request.Builder().url(server.url("/")).build());
+    Response response1 = call1.execute();
+    assertEquals(new X500Principal("CN=localhost"), response1.handshake().peerPrincipal());
+    assertEquals(null, response1.handshake().localPrincipal());
+    assertEquals("abc", response1.body().string());
+  }
+
+  @Test public void missingClientAuthFailsForNeeds() throws Exception {
+    OkHttpClient client = buildClient(null, intermediateCa);
+
+    SSLSocketFactory socketFactory = buildServerSslSocketFactory(ClientAuth.NEEDS);
+
+    server.useHttps(socketFactory, false);
+
+    Call call1 = client.newCall(new Request.Builder().url(server.url("/")).build());
+
+    try {
+      call1.execute();
+      fail();
+    } catch (ConnectException ce) {
+    }
+  }
+
+  @Test public void invalidClientAuthFails() throws Throwable {
+    HeldCertificate clientCert2 = new HeldCertificate.Builder()
+        .serialNumber("4")
+        .commonName("Jethro Willis")
+        .build();
+
+    OkHttpClient client = buildClient(clientCert2);
+
+    SSLSocketFactory socketFactory = buildServerSslSocketFactory(ClientAuth.NEEDS);
+
+    server.useHttps(socketFactory, false);
+
+    Call call1 = client.newCall(new Request.Builder().url(server.url("/")).build());
+
+    try {
+      call1.execute();
+      fail();
+    } catch (ConnectException ce) {
+    }
+  }
+
+  public OkHttpClient buildClient(HeldCertificate cert, HeldCertificate... chain) {
+    SslClient.Builder sslClientBuilder = new SslClient.Builder()
+        .addTrustedCertificate(rootCa.certificate);
+
+    if (cert != null) {
+      sslClientBuilder.certificateChain(cert, chain);
```

Comment:
off topic: should we change `SslClient` to use better words here? Currently it calls this `serverCert` but really it means `localCert`